[text structure] Find field and message structure endpoints (#105660)
elastic/elasticsearch: docs/reference/text-structure/apis/find-field-structure.asciidoc

[role="xpack"]
[[find-field-structure]]
= Find field structure API

Finds the structure of a field in an Elasticsearch index.

[discrete]
[[find-field-structure-request]]
== {api-request-title}

`GET _text_structure/find_field_structure`

[discrete]
[[find-field-structure-prereqs]]
== {api-prereq-title}

* If the {es} {security-features} are enabled, you must have `monitor_text_structure` or
`monitor` cluster privileges to use this API. See
<<security-privileges>>.

[discrete]
[[find-field-structure-desc]]
== {api-description-title}

> This API provides a starting point for ingesting data into {es} in a format that
> is suitable for subsequent use with other {stack} functionality.

Reviewer: ```suggestion
This API provides a starting point for extracting further information from log messages
already ingested into {es}. For example, if you have ingested data into a very simple
index that has just `@timestamp` and `message` fields, you can use this API to
see what common structure exists in the `message` field.
```

The response from the API contains:

* A couple of sample messages.
* Statistics that reveal the most common values for all fields detected within
the text and basic numeric statistics for numeric fields.
* Information about the structure of the text, which is useful when you write
ingest configurations to index it or similarly formatted text.
* Appropriate mappings for an {es} index, which you could use to ingest the text.

All this information can be calculated by the structure finder with no guidance.
However, you can optionally override some of the decisions about the text
structure by specifying one or more query parameters.

Details of the output can be seen in the <<find-field-structure-examples,examples>>.

If the structure finder produces unexpected results for some text,
specify the `explain` query parameter. It causes an `explanation` to appear in
the response, which should help in determining why the returned structure was
chosen.

[discrete]
[[find-field-structure-query-parms]]
== {api-query-parms-title}

`index`::

> (Required, string) The name of the index containing the field

Reviewer: ```suggestion
(Required, string) The name of the index containing the field.
```

include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-delimiter]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-explain]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-format]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-grok-pattern]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-ecs-compatibility]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-quote]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-should-trim-fields]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-timeout]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-timestamp-field]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-timestamp-format]

[discrete]
[[find-field-structure-examples]]
== {api-examples-title}

[discrete]
[[find-field-structure-example]]
=== Analyzing Elasticsearch log files

> Suppose you have a list of Elasticsearch logs messages in an index.

Reviewer: ```suggestion
Suppose you have a list of {es} log messages in an index.
```

You can analyze them with the `find_field_structure` endpoint as follows:

[source,console]
----
POST _bulk?refresh=true
{"index":{"_index":"test-logs"}}
{"message":"[2024-03-05T10:52:36,256][INFO ][o.a.l.u.VectorUtilPanamaProvider] [laptop] Java vector incubator API enabled; uses preferredBitSize=128"}
{"index":{"_index":"test-logs"}}
{"message":"[2024-03-05T10:52:41,038][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [repository-url]"}
{"index":{"_index":"test-logs"}}
{"message":"[2024-03-05T10:52:41,042][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [rest-root]"}
{"index":{"_index":"test-logs"}}

> {"message":"[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-core]tive methods for [MacOS]"}
> {"index":{"_index":"test-logs"}}
> {"message":"[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-redact]API enabled; uses preferredBitSize=128"}

Reviewer: It looks like this was created from one set of logs printed on top of another set.

```suggestion
{"message":"[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-core]"}
{"index":{"_index":"test-logs"}}
{"message":"[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-redact]"}
```

Author: very sharp... no clue how that happened 🤷

          "timezone" : "{{ event.timezone }}",
          "formats" : [
            "ISO8601"
          ]
        }
      },
      {
        "remove" : {
          "field" : "timestamp"
        }
      }
    ]
  },
  "field_stats" : {
    "loglevel" : {
      "count" : 22,
      "cardinality" : 1,
      "top_hits" : [
        {
          "value" : "INFO",
          "count" : 22
        }
      ]
    },
    "message" : {
      "count" : 22,
      "cardinality" : 22,
      "top_hits" : [
        {
          "value" : "[2024-03-05T10:52:36,256][INFO ][o.a.l.u.VectorUtilPanamaProvider] [laptop] Java vector incubator API enabled; uses preferredBitSize=128",
          "count" : 1
        },
        {
          "value" : "[2024-03-05T10:52:41,038][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [repository-url]",
          "count" : 1
        },
        {
          "value" : "[2024-03-05T10:52:41,042][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [rest-root]",
          "count" : 1
        },
        {
          "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [ingest-user-agent]",
          "count" : 1
        },
        {

>           "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-core]tive methods for [MacOS]",
>           "count" : 1
>         },
>         {
>           "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-redact]API enabled; uses preferredBitSize=128",

Reviewer: ```suggestion
          "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-core]",
          "count" : 1
        },
        {
          "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-redact]",
```

      ]
    },
    "timestamp" : {
      "count" : 22,
      "cardinality" : 14,
      "earliest" : "2024-03-05T10:52:36,256",
      "latest" : "2024-03-05T10:52:49,199",
      "top_hits" : [
        {
          "value" : "2024-03-05T10:52:41,044",
          "count" : 6
        },
        {
          "value" : "2024-03-05T10:52:41,043",
          "count" : 3
        },
        {
          "value" : "2024-03-05T10:52:41,059",
          "count" : 2
        },
        {
          "value" : "2024-03-05T10:52:36,256",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:41,038",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:41,042",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:43,291",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:46,098",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:47,227",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:47,259",
          "count" : 1
        }
      ]
    }
  }
}
----
// TESTRESPONSE[s/"sample_start" : ".*",/"sample_start" : "$body.sample_start",/]

> // The substitution is because the text is pre-processed by the test harness,
> // so the fields may get reordered in the JSON the endpoint sees

Reviewer: ```suggestion
// The substitution is because we cannot be sure of the order the
// documents will be retrieved from the input index
```

I would suggest running this test 5+ times before merging to see if there are any other fields that may change from run to run and cause subsequent CI failures. If there are any then they should be substituted too.

For a detailed description of the response format, or for additional examples

> on ingesting delimited text (such as CSV) or newline-delimited JSON, see the

Reviewer 2: Minor suggestion. Usually, we use `refer to` instead of `see` due to accessibility reasons. 
```suggestion
on ingesting delimited text (such as CSV) or newline-delimited JSON, refer to the
```
